allennlp/training/trainer.py

```diff
@@ -145,136 +161,243 @@ def train(self) -> None:
+        train_loss = 0.0
+        # Set the model to "train" mode.
+        self._model.train()
+
+        # Get tqdm for the training batches
+        train_generator = self._iterator(self._train_dataset, num_epochs=1)
         num_training_batches = self._iterator.get_num_batches(self._train_dataset)
+        train_generator_tqdm = tqdm.tqdm(train_generator,
+                                         disable=self._no_tqdm,
+                                         total=num_training_batches)
+        self._last_log = time.time()
+        batch_num = 0
+
+        logger.info("Training")
+        for batch in train_generator_tqdm:
+            # Increment batch and zero gradient
+            batch_num += 1
+            self._optimizer.zero_grad()
+
+            # Compute loss for batch and backpropagate
+            loss = self._compute_loss(batch, for_training=True)
+
+            # Make sure Variable is on the cpu before converting to numpy.
+            # .cpu() is a no-op if you aren't using GPUs.
+            train_loss += loss.data.cpu().numpy()
+
+            # Rescale gradients
+            self._rescale_gradients()
+
+            # Advance the optimizer
+            self._optimizer.step()
+
+            # Update the description with the latest metrics
+            metrics = self._get_metrics_for_batch(train_loss, batch_num)
+            description = self._description_from_metrics(metrics)
+            train_generator_tqdm.set_description(description)
+
+            # Log parameter values to Tensorboard
+            batch_num_total = num_training_batches * epoch + batch_num
+            if batch_num_total % self._summary_interval == 0:
+                for name, param in self._model.named_parameters():
+                    self._writer.add_train_scalar("parameter_mean/" + name, param.data.mean(), batch_num_total)
+                    self._writer.add_train_scalar("parameter_std/" + name, param.data.std(), batch_num_total)
+                    if param.grad is not None:
+                        self._writer.add_train_scalar("gradient_mean/" + name,
+                                                      param.grad.data.mean(),
+                                                      batch_num_total)
+                        self._writer.add_train_scalar("gradient_std/" + name,
+                                                      param.grad.data.std(),
+                                                      batch_num_total)
+                self._writer.add_train_scalar("loss/loss_train", metrics["loss"], batch_num_total)
+
+            # Log progress in no-tqdm case
+            if self._no_tqdm and time.time() - self._last_log > self._log_interval:
+                logger.info("Batch %d/%d: %s", batch_num, num_training_batches, description)
+                self._last_log = time.time()
+
+        # Return the metrics (and reset them)
+        return self._get_metrics_for_batch(train_loss, batch_num, reset=True)
```

Comment:
This is getting called twice, looks like.

Reply:
are you sure? the previous call is per-batch inside the loop, this is at the end of the loop to reset the metrics. or am I missing one?

Reply:
Ah, yeah, you're right.  I think the name referring to `for_batch` threw me off.  Not sure how to make the name better, though...

Reply:
I renamed to make this clearer